### PR TITLE
[Web][SIMS #1198] Restriction UI Updates

### PR DESCRIPTION
**Pull request by dheepak-aot**
## Student Restrictions UI updates

_Introduced a full page component for student called **student-page-container** which is extended from a full-page-container which will have alert banners for restrictions and SIN validation. 
All the student views are expected to use student-page-container instead of full-page-container to have the alert banners out of box.
Changing all the student views to use student page container is outside the scope of this story. But will do that part in a separate PR._

- [x] When a student has a notification_type (Warning) restriction (Regardless of type(federal/provincial)) display the Warning banner
image

- [x] When a student has a notification_type (Error) restriction (Regardless of type(federal/provincial)) display the Error banner
- [x] On click of "View activity" navigate to the "Account activity" page

image

- [x] Create new button "Account activity" under the user icon on the top right for a student

image

- [x] On click of "Account activity" display the "Account Activity" page
- [x] Display the restrictions that currently exist for a student that have notification_type of (Warning or Error)
- [x] Display the content per figma (the rest will be static content, descriptors under each drop down is in the "notification" on the spreadsheet

- [x] If a student attempts to submit an application with 1 of the above restrictions display toast message "you have a restriction on account" (reuse the same restriction toast message currently in place)
image

### CODE REVIEW

File: sources/packages/web/src/assets/css/formio-shared.scss [hunk 1]
```diff
@@ -438,3 +438,15 @@
   border: 1px solid $common-border-color !important;
   border-radius: 8px !important;
 }
+
+//Style added hide the header of data grid for student activity page.
+.student-activity-grid th {
+  border: none !important;
```
Comment (ann-aot) [inline sources/packages/web/src/assets/css/formio-shared.scss:444]: is this only to remove the border and header? then `formio-datagrid-unset` itself should work, unless I am missing something
image

Comment (dheepak-aot) [reply sources/packages/web/src/assets/css/formio-shared.scss]: You are right. I missed to remove this part from .scss
My form uses formio-datagrid-unset.

student-activity-grid used just to left-pad the item

File: sources/packages/forms/src/form-definitions/studentaccountactivity.json [hunk 2]
```diff
@@ -0,0 +1,10081 @@
...
+                  "tooltip": "",
+                  "disableAddingRemovingRows": true,
+                  "conditionalAddButton": "",
+                  "reorder": false,
+                  "addAnother": "",
+                  "addAnotherPosition": "bottom",
+                  "layoutFixed": false,
+                  "enableRowGroups": false,
+                  "initEmpty": false,
+                  "customClass": "formio-datagrid-unset student-activity-grid mt-n2",
```
Comment (andrewsignori-aot) [inline sources/packages/forms/src/form-definitions/studentaccountactivity.json:226]: Were you able to reduce the spacing between the list items? I noticed that the print on the PR description has only one item. How does it look when there are 2 items.

File: sources/packages/forms/src/form-definitions/studentaccountactivity.json [hunk 3]
```diff
@@ -0,0 +1,10081 @@
...
+              "customClass": "",
+              "collapsible": false,
+              "hidden": false,
+              "hideLabel": true,
+              "disabled": false,
+              "modalEdit": false,
+              "key": "panel1",
+              "tags": [],
+              "properties": {},
+              "customConditional": "show = !!(data.restrictions && data.restrictions.length);",
```
[the remark this answers is not in the document]
Comment (dheepak-aot) [reply sources/packages/forms/src/form-definitions/studentaccountactivity.json:31]: I missed to remove in the first panel. Will remove it. 
image

File: sources/packages/web/src/composables/useStudentStore.ts [hunk 4]
```diff
@@ -20,10 +21,35 @@ export function useStudentStore(rootStore?: Store<any>) {
...
+  const hasWarning = computed<boolean>(() => {
+    return store.getters["student/hasWarning"];
+  });
+
+  const getRestrictions = (): StudentRestriction[] => {
+    return store.getters["student/getRestrictions"];
+  };
+
+  const activeRestrictions = computed<StudentRestriction[]>(() => {
+    return store.getters["student/getRestrictions"];
```
[the remark this answers is not in the document]
Comment (dheepak-aot) [reply sources/packages/web/src/composables/useStudentStore.ts:41]: getRestrictions() needs to deleted. Will remove it.

File: sources/packages/web/src/assets/css/formio-shared.scss [hunk 5]
```diff
@@ -438,3 +438,8 @@
   border: 1px solid $common-border-color !important;
   border-radius: 8px !important;
 }
+
+//Style to add padding to the bullet item.
+.formio-datagrid-unset.student-activity-grid .table-bordered td {
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/assets/css/formio-shared.scss:443]: Is there a way to this be more generic or there is no need to reuse it anywhere else?

File: sources/packages/web/src/components/generic/Banner.vue [hunk 6]
```diff
@@ -1,20 +1,20 @@
...
-      <v-col cols="2"><slot name="actions"></slot></v-col>
+      <v-col sm="2" md="2" lg="2"><slot name="actions"></slot></v-col>
     </v-row>
   </v-alert>
 </template>
 <script lang="ts">
 export default {
   props: {
-    color: {
+    type: {
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/generic/Banner.vue:17]: A validator can be added to the type. We have a couple of similar properties doing it already, like below:
```typescript
validator: (val: string) => val in LayoutTemplates
```

File: sources/packages/forms/src/form-definitions/studentaccountactivity.json (hunk 2 above)
Comment (dheepak-aot) [reply sources/packages/forms/src/form-definitions/studentaccountactivity.json:226]: I did not try further to reduce the spacing between the items. Used the existing class which was present. Added and additional style for bullet left padding. 
image

File: sources/packages/web/src/components/generic/Banner.models.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+export enum BannerTypes {
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/generic/Banner.models.ts:1]: Do we need to consume this outside the component?

File: sources/packages/web/src/assets/css/formio-shared.scss (hunk 5 above)
Comment (dheepak-aot) [reply sources/packages/web/src/assets/css/formio-shared.scss:443]: It is something which can potentially come again. So will try to make it generic.

File: sources/packages/forms/src/form-definitions/studentaccountactivity.json [hunk 8]
```diff
@@ -0,0 +1,10081 @@
...
+                  },
+                  "type": "panel",
+                  "label": "Panel",
+                  "breadcrumb": "default",
+                  "tabindex": "",
+                  "input": false,
+                  "tableView": false,
+                  "components": [
+                    {
+                      "html": "<p>The&nbsp;Canada Student Financial Assistance Program (CSFA Program) has reported an issue with your repayment assistance program. If you need more information, please contact the National Student Loan Service Centre.</p>",
```
Comment (ann-aot) [inline sources/packages/forms/src/form-definitions/studentaccountactivity.json:6442]: why `&nbsp`?? we can use bootstrap classes. it's in 26 places

File: sources/packages/web/src/components/layouts/student/StudentPageContainer.vue [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+      <restriction-banner v-if="!hideRestriction" />
+      <CheckValidSINBanner />
+    </template>
+    <slot name="content"></slot>
+  </full-page-container>
+</template>
+<script lang="ts">
+import RestrictionBanner from "@/views/student/RestrictionBanner.vue";
+import CheckValidSINBanner from "@/views/student/CheckValidSINBanner.vue";
+enum LayoutTemplates {
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/layouts/student/StudentPageContainer.vue:19]: Should we duplicate this? Maybe it can be moved to the contracts folder.

File: sources/packages/forms/src/form-definitions/studentaccountactivity.json [hunk 10]
```diff
@@ -0,0 +1,10081 @@
...
+                      "allowMultipleMasks": false,
+                      "addons": [],
+                      "id": "el0otc"
+                    }
+                  ],
+                  "placeholder": "",
+                  "prefix": "",
+                  "suffix": "",
+                  "multiple": false,
+
```
Comment (ann-aot) [inline sources/packages/forms/src/form-definitions/studentaccountactivity.json:7958]: do we  need `&nbsp` here?

File: sources/packages/forms/src/form-definitions/studentaccountactivity.json (hunk 8 above)
Comment (dheepak-aot) [reply sources/packages/forms/src/form-definitions/studentaccountactivity.json:6442]: I never added the &nbsp. The most part of this form was done by @abschwenker.

File: sources/packages/web/src/composables/useStudentStore.ts [hunk 11]
```diff
@@ -20,10 +21,35 @@ export function useStudentStore(rootStore?: Store<any>) {
...
 
+  const updateRestrictions = async () => {
+    await store.dispatch("student/updateStudentRestrictions");
+  };
+
+  const hasRestriction = computed<boolean>(() => {
+    return store.getters["student/hasRestriction"];
+  });
+
+  const hasWarning = computed<boolean>(() => {
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/composables/useStudentStore.ts:32]: Maybe we can be more specific here and name it as `hasErrorRestriction`.

File: sources/packages/web/src/composables/useStudentStore.ts [hunk 12]
```diff
@@ -20,10 +21,35 @@ export function useStudentStore(rootStore?: Store<any>) {
     await store.dispatch("student/setHasStudentAccount", hasAccount);
   };
 
+  const updateRestrictions = async () => {
+    await store.dispatch("student/updateStudentRestrictions");
+  };
+
+  const hasRestriction = computed<boolean>(() => {
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/composables/useStudentStore.ts:28]: Maybe we can be more specific here and name it as `hasRestrictionError`.

File: sources/packages/web/src/composables/useStudentStore.ts [hunk 13]
```diff
@@ -20,10 +21,35 @@ export function useStudentStore(rootStore?: Store<any>) {
...
+    return store.getters["student/hasRestriction"];
+  });
+
+  const hasWarning = computed<boolean>(() => {
+    return store.getters["student/hasWarning"];
+  });
+
+  const getRestrictions = (): StudentRestriction[] => {
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/composables/useStudentStore.ts:36]: Please convert this to a computed with a proper name.

File: sources/packages/web/src/store/modules/student/actions.ts [hunk 14]
```diff
@@ -17,4 +17,11 @@ export const actions: ActionTree<StudentState, RootState> = {
   ) {
     context.commit("setHasStudentAccount", hasStudentAccount);
   },
+
+  async updateStudentRestrictions(
+    context: ActionContext<StudentState, RootState>,
+  ) {
+    const restrictions = await StudentService.shared.getStudentRestriction();
+    context.commit("updateStudentRestrictions", restrictions);
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/store/modules/student/actions.ts:25]: I noticed that the interface `StudentRestriction` was created but we are actually saving to the state the DTO, it is the expected behavior?

File: sources/packages/web/src/components/generic/Banner.models.ts (hunk 7 above)
Comment (dheepak-aot) [reply sources/packages/web/src/components/generic/Banner.models.ts:1]: Yes. we use this enum to set the attribute while using the component.

File: sources/packages/web/src/store/modules/student/getters.ts [hunk 15]
```diff
@@ -1,7 +1,23 @@
...
+  RootState,
+  RestrictionNotificationType,
+  StudentRestriction,
+} from "@/types";
 import { StudentState } from "./student";
 
 export const getters: GetterTree<StudentState, RootState> = {
-  // To be implemented if needed.
+  hasRestriction(state: StudentState): boolean {
+    return state.restrictions?.some(
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/store/modules/student/getters.ts:11]: We can ensure that the state will always have an empty array and avoid the null checks every time.

File: sources/packages/web/src/types/contracts/StudentContract.ts [hunk 16]
```diff
@@ -33,3 +34,17 @@ export enum FileOriginType {
   Student = "Student",
   Ministry = "Ministry",
 }
+
+/**
+ * Contract for student restriction.
+ */
+export interface StudentRestriction {
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/types/contracts/StudentContract.ts:41]: Is this a DTO? If it is used only for the state, please place it alongside the other state types.

File: sources/packages/web/src/views/institution/locations/programs/LocationProgramAddEdit.vue [hunk 17]
```diff
@@ -41,7 +41,7 @@ import { InstitutionService } from "@/services/InstitutionService";
 import { ClientIdType } from "@/types";
 import { useToastMessage } from "@/composables";
 import { AuthService } from "@/services/AuthService";
-
+import { BannerTypes } from "@/components/generic/Banner.models";
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/institution/locations/programs/LocationProgramAddEdit.vue:44]: Do we need to set this property using an enum?

File: sources/packages/web/src/views/student/RestrictionBanner.vue [hunk 18]
```diff
@@ -1,14 +1,59 @@
...
-  </Message>
+  <banner
+    class="mb-2"
+    v-if="restriction"
+    :type="BannerTypes.Error"
+    header="Your account has a restriction"
+    summary="You have a restriction on your account that must be resolved to continue with your student financial aid. Please view the message and resolve the items to minimize disruption and impact."
+  >
+    <template #actions>
+      <v-btn :color="BannerTypes.Error" @click="viewStudentAccountActivity">
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/student/RestrictionBanner.vue:10]: `<v-btn :color="BannerTypes.Error"` this does not seem right? Either we need a more generic enum or we could use just the strings for this case, which seem to be the Vue way of doing things.

File: sources/packages/web/src/views/student/RestrictionBanner.vue [hunk 19]
```diff
@@ -1,14 +1,59 @@
...
+    const router = useRouter();
+    const { updateRestrictions, hasRestriction, hasWarning } =
+      useStudentStore();
+
+    const restriction = computed(() => {
+      return hasRestriction.value;
+    });
+
+    const warning = computed(() => {
+      return hasWarning.value;
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/student/RestrictionBanner.vue:45]: `hasWarning` can be used directly, no need to create another computed.

File: sources/packages/web/src/views/student/CheckValidSINBanner.vue [hunk 20]
```diff
@@ -2,7 +2,7 @@
   <banner
```
Comment (ann-aot) [inline sources/packages/web/src/views/student/CheckValidSINBanner.vue:2]: i know its not part of the PR, this is not a  view it a component, it should be `component` folder

File: sources/packages/web/src/views/student/RestrictionBanner.vue [hunk 21]
```diff
@@ -1,14 +1,59 @@
 <template>
-  <Message severity="error">
-    {{ restrictionMessage }}
-  </Message>
+  <banner
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/student/RestrictionBanner.vue:2]: `RestrictionBanner.vue` is a component, right? I believe that it must be placed inside the components.

File: sources/packages/web/src/views/student/StudentAccountActivityFormModel.ts [hunk 22]
```diff
@@ -0,0 +1,4 @@
+import { StudentRestriction } from "@/types";
+export interface StudentAccountActivityFormModel {
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/student/StudentAccountActivityFormModel.ts:2]: Is it something that will be used anywhere else? If not, I believe that it can be defined inside the component.

File: sources/packages/web/src/views/student/StudentAccountActivity.vue [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+            restriction.type === RestrictionNotificationType.Error,
+        ),
+      };
+    });
+    onMounted(async () => {
+      await updateRestrictions();
+    });
+    return {
+      StudentRoutesConst,
+      activeRestrictions,
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/student/StudentAccountActivity.vue:43]: I believe that there is no reason to return the `activeRestrictions`.

File: sources/packages/web/src/views/student/StudentApplicationSummary.vue [hunk 24]
```diff
@@ -37,30 +31,25 @@
...
 import { ApplicationStatus } from "@/types";
 import StudentApplications from "@/components/aest/StudentApplications.vue";
-import CheckValidSINBanner from "@/views/student/CheckValidSINBanner.vue";
 import { ApplicationService } from "@/services/ApplicationService";
 import { useRouter } from "vue-router";
 import { StudentRoutesConst } from "@/constants/routes/RouteConstants";
 import { useToastMessage, ModalDialog } from "@/composables";
 import ConfirmEditApplication from "@/components/students/modals/ConfirmEditApplication.vue";
 import CancelApplication from "@/components/students/modals/CancelApplicationModal.vue";
+import StudentPageContainer from "@/components/layouts/student/StudentPageContainer.vue";
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/student/StudentApplicationSummary.vue:42]: This component was added to the main.ts, right?

**Review by andrewsignori-aot — COMMENTED** [review]
Great to see the simplifications around restrictions, nice work. Please take a look at the comments.
For the below ACs, what is the idea? The student will receive the notification only upon application submission, right?
image

File: sources/packages/web/src/views/student/StudentProfile.vue [hunk 25]
```diff
@@ -4,10 +4,7 @@
       <header-navigator v-if="editMode" title="Student" subTitle="Profile" />
     </template>
     <template #alerts>
-      <RestrictionBanner
-        v-if="hasRestriction"
-        :restrictionMessage="restrictionMessage"
-      />
+      <restriction-banner />
```
Comment (ann-aot) [inline sources/packages/web/src/views/student/StudentProfile.vue:7]: we need to use `StudentPageContainer` for all student views right??

File: sources/packages/web/src/composables/useStudentStore.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (andrewsignori-aot) [reply sources/packages/web/src/composables/useStudentStore.ts:41]: I added the same comment, please disregard mine 😉

File: sources/packages/web/src/components/layouts/student/StudentPageContainer.vue (hunk 9 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/components/layouts/student/StudentPageContainer.vue:19]: This is an example where the enum is present but we use the strings to set the values. So the enum was created for the internal component organization, but the component consumer can provide the value as strings, the same behavior for other Vue components in general (I believe).

File: sources/packages/web/src/composables/useStudentStore.ts (hunk 13 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/composables/useStudentStore.ts:36]: I saw another comment about the same thing, hence I am closing this one.

File: sources/packages/web/src/types/contracts/RestrictionContract.ts [hunk 26]
```diff
@@ -68,3 +68,21 @@ export enum RestrictionEntityType {
...
+  NoEffect = "No effect",
+  /**
+   * Warning.
+   */
+  Warning = "Warning",
+  /**
+   * Error.
+   */
+  Error = "Error",
+}
```
Comment (henryoforeh-aot) [inline sources/packages/web/src/types/contracts/RestrictionContract.ts:88]: The assigned constants here begin with Upper case while in this other file it begins with lower case. Is there any pattern we are adhering to ? 
image

File: sources/packages/web/src/views/student/StudentApplicationDetails.vue [hunk 27]
```diff
@@ -1,9 +1,6 @@
 <template>
   <div class="p-m-4">
-    <RestrictionBanner
-      v-if="hasRestriction"
-      :restrictionMessage="restrictionMessage"
-    />
+    <restriction-banner />
```
[the remark this answers is not in the document]
Comment (dheepak-aot) [reply sources/packages/web/src/views/student/StudentApplicationDetails.vue:3]: Sorry I must have updated the PR description precisely. Changing all the student views to use student page container is outside the scope of this story. Will create a technical debt for the same.

File: sources/packages/web/src/components/layouts/student/StudentPageContainer.vue (hunk 9 above)
Comment (guru-aot) [reply sources/packages/web/src/components/layouts/student/StudentPageContainer.vue]: I believe, we can have it in the template rather having it in the consumer everytime.

File: sources/packages/web/src/types/contracts/RestrictionContract.ts (hunk 26 above)
Comment (dheepak-aot) [reply sources/packages/web/src/types/contracts/RestrictionContract.ts:88]: We are adhering to a pattern that all the enum property names must start with uppercase.  And the key and value not required to Match case and value in some cases.
https://github.com/bcgov/SIMS/wiki/Development-Standards-and-Decisions

File: sources/packages/web/src/views/student/StudentApplicationDetails.vue (hunk 27 above)
[the remark this answers is not in the document]
Comment (dheepak-aot) [reply sources/packages/web/src/views/student/StudentApplicationDetails.vue:3]: Correction to my previous statement. I will be doing a separate PR in the story to change all the student views to student-page-container.

**Review by guru-aot — APPROVED** [review]
LGTM, Nice work @dheepak-aot

File: sources/packages/web/src/components/layouts/student/StudentPageContainer.vue (hunk 9 above)
Comment (ann-aot) [reply sources/packages/web/src/components/layouts/student/StudentPageContainer.vue]: mm... Iam okay with both

Comment (andrewsignori-aot) [reply sources/packages/web/src/components/layouts/student/StudentPageContainer.vue]: Trying to settle this, I believe that we can have both ways we just do not need to enforce it.

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for doing the changes, looks good 👍

**Review by ann-aot — APPROVED** [review]
Thanks for doing the changes 👍 Good Work @dheepak-aot
